src/txn/txn.c

```diff
@@ -864,8 +866,30 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
 			    op->type != WT_TXN_OP_BASIC_TS) {
 				WT_ASSERT(session,
 				    op->fileid != WT_METAFILE_ID);
-				__wt_timestamp_set(&op->u.upd->timestamp,
-				    &txn->commit_timestamp);
+
+				/*
+				 * If this commit is after the prepare,
+				 * then need to update the timestamp as
+				 * commit timestamp. Updating timestamp
+				 * might not be atomic operation, hence
+				 * we will lock the update structure for
+				 * read using the state as LOCKED.
+				 */
+				if (FLD_ISSET(op->u.upd->state,
+				    WT_UPDATE_PREPARE)) {
+					FLD_SET(op->u.upd->state,
+					    WT_UPDATE_LOCKED);
+
+					__wt_timestamp_set(
+					    &op->u.upd->timestamp,
+					    &txn->commit_timestamp);
+
+					FLD_SET(op->u.upd->state,
+					    WT_UPDATE_NONE);
```

Comment:
Does the unlock need to happen after all updates have had their timestamps reset?

Reply:
Agreed in discussion that this unlock is ok, but need to make this logic as a function call.